### DISCUSSION TITLE
Request for assistance configuring WebRTC Camera device

### DISCUSSION THREAD

**Discussion by bkdfkup** [mikejac/node-red-contrib-google-smarthome#459] [General]
as logged on node-red forums here https://discourse.nodered.org/t/camera-configuration-using-node-red-contrib-google-smarthome/57545 i am seeking assistance regarding configuration of a WebRTC stream in the google device camera configuration. Some detailed guidance would be appreciated if possible.

**Comment by bkdfkup** [comment]
Excerpt from node-red forum post.

----------------------------------------------------------------------------------

Hi All.

Working on configuring camera streaming to google assistant via
node-red-contrib-google-smarthome and am having some challenging issues. I have tried several protocols, and have only had some success using FFMPEG-produced smooth-streaming. No luck using MP4, HLS, DASH or WebRTC protocols.

I am attempting to get WebRTC working now as my attempts to live stream an rtsp camera feed via FFMPEG were causing some issues.

I have configured a functioning instance of OvenMediaEngine, and have a working WebRTC stream being produced. What i cannot now figure out are the correct WebRTC configurations to enter into the google device node configurations.

Using the example configurations here[ WebRTC Streaming - OvenMediaEngine](https://airensoft.gitbook.io/ovenmediaengine/streaming/webrtc-publishing) and here [Getting Started - OvenMediaEngine](https://airensoft.gitbook.io/ovenmediaengine/getting-started) i would love some help determining what needs to be entered here
image

Thanks in advance

![Screenshot 2022-01-31 222046](https://user-images.githubusercontent.com/74683396/151784991-e806f3bc-5fb4-4b15-bccf-fb2594ac5a54.png)

**Comment by ckhmer1** [comment]
I've developed the WebRTC camera support, I'm sorry but I don't own a WebRTC camera, the feature is untested by me.
I can provide You with limited support.

**Comment by bkdfkup** [comment]
Thanks for replying. The links I sent you may be of benefit to you for testing purposes. It can produce many stream types including WebRTC and DASH.

What protocols have you tested successfully so I can prioritise which I use?
________________________________

**Comment by ckhmer1** [comment]
Unfortunately I don't own an in IP camera, I'm trying to simulate then with an usb camera and nginx/rtmp, currently without success.
I've tested the node only simulating the messages.

**Comment by bkdfkup** [comment]
Understood! Thank you so much for your work on the node in general. It has opened up so many possibilities.

One suggestion is for you to use a public rtmp stream and re-encode that to whatever format. If I make any progress I'll let you know. 👍
________________________________

**Comment by nostroff** [comment]
I've had partial success. The good news is that node-red-contrib-google-smarthome module appears to work fine. The settings I used are as follows:
WebRTC Signaling URL: https://[my public server.com]:9443/stream/demo/channel/0/webrtc?uuid=demo&channel=0
WebRTC Offer session description protocol (SDP): o=- 6910163479322986733 2 IN IP4 127.0.0.1
WebRTC IceServers (entered as JSON, not string): [{"urls":"stun:stun.l.google.com:19302"}]

From the log messages, it looks like the public server needs to have a valid SSL certificate.

My camera streams video using RTSP. To convert the RTSP to WebRTC and provide the signaling server, I installed RTSPtoWeb (see: https://github.com/deepch/RTSPtoWeb). 

I suspect my roadblock is with my camera not supporting the correct video and/or audio codec for playback on a Google Nest Hub 2 since I'm able to successfully stream my camera via WebRTC to a Chrome browser by browsing to https://[my public server.com]:9443. Some information regarding compatible video/audio codecs is here: https://developers.google.com/cast/docs/media. 

Good luck.

**Comment by ckhmer1** [comment]
Hello,
i achieved similar results using the chromecast and simulating the webcam with nginx.
I also suspect that the codecs are not the right ones.
Thanks for the feedback.

**Comment by Caprico85** [comment]
Converting this to a discussion to close this issue but keep the information available for people looking for help.
